src/ripple/conditions/RsaSha256.h

```diff
@@ -0,0 +1,319 @@
+#include <ed25519-donna/ed25519.h>
+#include <openssl/bn.h>
+#include <openssl/evp.h>
+#include <openssl/rsa.h>
+#include <openssl/pem.h>
+#include <openssl/sha.h>
+
+#include <boost/algorithm/clamp.hpp>
+#include <boost/regex.hpp>
+#include <boost/optional.hpp>
+#include <algorithm>
+#include <iterator>
+
+namespace ripple {
+namespace cryptoconditions {
+
+class RsaSha256 final
+    : public Fulfillment
+{
+    struct rsa_deleter
+    {
+        void operator() (RSA* rsa) const
+        {
+            RSA_free (rsa);
+        }
+    };
+
+    using RsaKey = std::unique_ptr<RSA, rsa_deleter>;
+
+    struct bn_deleter
+    {
+        void operator() (BIGNUM* bn) const
+        {
+            BN_free (bn);
+        }
+    };
+
+    using BigNum = std::unique_ptr<BIGNUM, bn_deleter>;
+
+    Buffer modulus_;
+    Buffer signature_;
+
+    // Check whether the public modulus meets the length
+    // requirements imposed by section 4.4.1 of the RFC.
+    static
+    bool
+    checkModulusLength (int len)
+    {
+        if (len <= 0)
+            return false;
+
+        return len == boost::algorithm::clamp(len, 128, 512);
+    }
+
+    bool sign_helper (RSA* key, Slice message)
+    {
+        int const keySize = RSA_size(key);
+        if (!checkModulusLength (keySize))
+            return false;
+
+        sha256_hasher h;
+        h (message.data(), message.size());
+        auto digest = static_cast<sha256_hasher::result_type>(h);
+
+        // Pad the result (-1 -> use hash length as salt length)
+        Buffer buf;
+
+        if (!RSA_padding_add_PKCS1_PSS(key,
+                buf.alloc(keySize), digest.data(),
+                EVP_sha256(), -1))
+            return false;
+
+        // Sign - we've manually padded the input already.
+        auto ret = RSA_private_encrypt(keySize, buf.data(),
+            signature_.alloc (buf.size()), key, RSA_NO_PADDING);
+
+        if (ret == -1)
+        {
+            signature_.clear();
+            return false;
+        }
+
+        BN_bn2bin (key->n, modulus_.alloc(BN_num_bytes (key->n)));
+        return true;
+    }
+
+    bool
+    validate_helper (RSA* key, Slice message) const
+    {
+        int const keySize = RSA_size(key);
+        if (!checkModulusLength (keySize))
+            return false;
+
+        Buffer buf;
+
+        auto ret = RSA_public_decrypt(
+            keySize,
+            signature_.data(),
+            buf.alloc (keySize),
+            key,
+            RSA_NO_PADDING);
+
+        if (ret == -1)
+            return false;
+
+        sha256_hasher h;
+        h (message.data(), message.size());
+        auto digest = static_cast<sha256_hasher::result_type>(h);
+
+        return RSA_verify_PKCS1_PSS(key, digest.data(), EVP_sha256(), buf.data(), -1) == 1;
+    }
+
+    bool parsePayloadHelper(Slice s)
+    {
+        auto iter = s.begin ();
+
+        std::size_t len;
+
+        std::tie (iter, len) = oer::decode_length (
+            iter, s.end());
```

Comment:
`decode_length` dereferences iter without a check. If this function can never take an empty Slice it's not a problem, otherwise we need to check for empty before this call.